File: backend/app/admin_settings/schemas.py
"""Schemas Pydantic para AdminSettings panel /admin/settings.

Validación shape JSONB ANTES de persistir. ENS-correctness:

- ``extra="forbid"`` en cada categoría: rechaza payloads con campos
  no esperados (defensa contra typos del frontend o intentos de
  poison data desde panel admin).
- Type-safe boundaries: tipos explícitos, sin defaults silenciosos
  para campos críticos.
- Pattern validation donde aplica (hex color, locale ISO, IANA
  timezone via str + service-layer check).

Distinción con `app.config.Settings`:

- `Settings`: env-based, immutable runtime, defaults conservadores
- `AdminSettings` (este módulo): BD-based, mutable runtime, override
  panel admin con audit trail (audit_log)

PATCH atómico por categoría (no endpoint "PATCH all"):

- PATCH /api/v1/admin/settings/branding       → BrandingSettings
- PATCH /api/v1/admin/settings/notifications  → NotificationsSettings
- PATCH /api/v1/admin/settings/smtp           → SmtpSettings
- PATCH /api/v1/admin/settings/general        → GeneralSettings
- PATCH /api/v1/admin/settings/analytics      → AnalyticsPrefs

Cada PATCH usa la category schema directamente. `exclude_unset=True`
en service layer permite parciales (update solo campos enviados).

Ver plan v4.2 sección FASE 4 + sub-bloque 4.A.2.a + ADR pendiente
si decisiones nuevas surgen.
"""
from __future__ import annotations

from datetime import datetime
from typing import Dict, List, Literal, Optional
from uuid import UUID

from pydantic import BaseModel, ConfigDict, EmailStr, Field


# ─────────────────────────────────────────────────────────────────
# Categoría: branding (logos, colores corporativos, footer text)
# ─────────────────────────────────────────────────────────────────


class BrandingSettings(BaseModel):
    """Identidad visual: logo, colores, footer text."""

    model_config = ConfigDict(extra="forbid")

    logo_url: Optional[str] = Field(default=None, max_length=500)
    primary_color: Optional[str] = Field(
        default=None,
        pattern=r"^#[0-9A-Fa-f]{6}$",
        description="Color hex format #RRGGBB",
    )
    secondary_color: Optional[str] = Field(
        default=None,
        pattern=r"^#[0-9A-Fa-f]{6}$",
        description="Color hex format #RRGGBB",
    )
    footer_text: Optional[str] = Field(default=None, max_length=500)


# ─────────────────────────────────────────────────────────────────
# Categoría: notifications (email forward + magic link config v4)
# Plan v4.2 sección 4.2.2 (código literal preservado)
# ─────────────────────────────────────────────────────────────────


class MagicLinkPurposeOverride(BaseModel):
    """Override por purpose en magic_link_per_purpose_overrides dict."""

    model_config = ConfigDict(extra="forbid")

    enabled: bool = True
    default_recipient_email: Optional[EmailStr] = None
    cc_emails: List[EmailStr] = Field(default_factory=list)
    template_id: Optional[str] = None


class NotificationsSettings(BaseModel):
    """Email forward + magic link sender config + digest preferences."""

    model_config = ConfigDict(extra="forbid")

    client_messages_forward_enabled: bool = True
    client_messages_forward_to: Optional[EmailStr] = None
    smtp_custom: Optional[Dict] = None
    digest_enabled: bool = True
    digest_time_local: str = Field(
        default="08:00",
        pattern=r"^([01]\d|2[0-3]):[0-5]\d$",
        description="Hora local HH:MM 24h format",
    )
    magic_link_default_sender: Optional[EmailStr] = None
    magic_link_per_purpose_overrides: Dict[str, MagicLinkPurposeOverride] = Field(
        default_factory=dict,
    )
    # #26 · número WhatsApp de Marcos editable desde el panel (antes solo env →
    # no se podía cambiar el destino de notificaciones inbound sin reiniciar).
    # Si vacío, el dispatcher cae al valor de entorno.
    marcos_whatsapp_number: Optional[str] = Field(
        default=None,
        description=(
            "Número WhatsApp de Marcos en E.164 (ej. +34637165328) · destino de "
            "notificaciones inbound. Vacío = usa el valor de entorno."
        ),
        pattern=r"^\+?[1-9]\d{6,14}$",
    )


# ─────────────────────────────────────────────────────────────────
# Categoría: smtp (override env-based defaults Settings)
# ─────────────────────────────────────────────────────────────────


class SmtpSettings(BaseModel):
    """SMTP override custom. Settings env-based son fallback."""

    model_config = ConfigDict(extra="forbid")

    host: Optional[str] = Field(default=None, max_length=255)
    port: Optional[int] = Field(default=None, ge=1, le=65535)
    username: Optional[str] = Field(default=None, max_length=255)
    password: Optional[str] = Field(
        default=None,
        max_length=500,
        description="Password en plano (BD-encrypted en sub-bloque posterior si se requiere)",
    )


# ─────────────────────────────────────────────────────────────────
# Categoría: general (timezone, locale, formato fecha)
# ─────────────────────────────────────────────────────────────────


class GeneralSettings(BaseModel):
    """Configuración general: timezone, locale, formato fecha."""

    model_config = ConfigDict(extra="forbid")

    timezone: str = Field(
        default="Europe/Madrid",
        max_length=50,
        description="IANA timezone identifier (validación contra zoneinfo en service layer)",
    )
    locale: str = Field(
        default="es-ES",
        pattern=r"^[a-z]{2}-[A-Z]{2}$",
        description="Locale ISO format ll-CC",
    )
    date_format: str = Field(
        default="DD/MM/YYYY",
        max_length=20,
    )


# ─────────────────────────────────────────────────────────────────
# Categoría: analytics_prefs (privacidad analytics)
# ─────────────────────────────────────────────────────────────────


class AnalyticsPrefs(BaseModel):
    """Preferencias analytics. Solo campo show_corpus_metric en plan v4.2."""

    model_config = ConfigDict(extra="forbid")

    show_corpus_metric: bool = True


# ─────────────────────────────────────────────────────────────────
# Categoría: fiscal (identidad fiscal del emisor/consultor · punto #44)
# FUENTE ÚNICA: NIF, nombre fiscal/comercial, tipo persona F/J, domicilio
# fiscal estructurado, régimen IVA/IRPF, datos bancarios. Consumida por
# QR Verifactu, PDF factura, contratos/DPA, RoPA, Facturae seller vía
# core.fiscal_identity.get_fiscal_identity. Default régimen: autónomo
# persona física (tipo_persona='F', sujeto_irpf=True).
# ─────────────────────────────────────────────────────────────────


class FiscalSettings(BaseModel):
    """Identidad fiscal del emisor. Todos los campos opcionales (PATCH
    parcial), pero los consumidores exigen al menos ``nif`` +
    ``nombre_fiscal`` + domicilio para emitir factura/contrato válidos."""

    model_config = ConfigDict(extra="forbid")

    # Identidad
    nif: Optional[str] = Field(
        default=None, max_length=20,
        description="NIF/CIF del emisor. Autónomo persona física: NIF con letra (ej. 77171140E).",
    )
    nombre_fiscal: Optional[str] = Field(
        default=None, max_length=200,
        description="Nombre/razón fiscal canónica (persona física del autónomo o razón social SL).",
    )
    nombre_comercial: Optional[str] = Field(
        default=None, max_length=200,
        description="Nombre comercial/marca (ej. FULKRO). Si vacío se usa nombre_fiscal.",
    )
    tipo_persona: Literal["F", "J"] = Field(
        default="F",
        description="F=persona física (autónomo) · J=jurídica (sociedad). Determina PersonTypeCode Facturae.",
    )

    # Domicilio fiscal estructurado (Facturae exige vía + CP + municipio + provincia)
    domicilio_via: Optional[str] = Field(default=None, max_length=250)
    domicilio_cp: Optional[str] = Field(default=None, max_length=10)
    domicilio_municipio: Optional[str] = Field(default=None, max_length=120)
    domicilio_provincia: Optional[str] = Field(default=None, max_length=120)
    domicilio_pais: str = Field(default="España", max_length=80)

    # Régimen impositivo
    iva_pct: float = Field(default=21.0, ge=0, le=100, description="Tipo IVA general aplicable (%).")
    sujeto_irpf: bool = Field(
        default=True,
        description="Autónomo: facturas a empresa/AAPP con retención IRPF.",
    )
    irpf_pct: float = Field(
        default=15.0, ge=0, le=100,
        description="% retención IRPF. Nuevo autónomo puede aplicar 7% reducido (3 primeros años).",
    )

    # Datos bancarios (fuente única · fallback Settings.marcos_bank_* en transición)
    iban: Optional[str] = Field(default=None, max_length=34)
    bank_holder: Optional[str] = Field(default=None, max_length=200)
    bank_institution: Optional[str] = Field(default=None, max_length=120)
    bank_bic: Optional[str] = Field(default=None, max_length=11)


# ─────────────────────────────────────────────────────────────────
# About endpoint (read-only stats)
# Plan v4.2 sección 4.2.3 (código literal preservado)
# ─────────────────────────────────────────────────────────────────


class AdminSettingsAbout(BaseModel):
    """Stats read-only para tab About: versión + corpus + suite."""

    version: str
    commit_hash: str
    uptime_days: int
    active_clients_count: int
    corpus_sources_count: int
    corpus_sources_target: int
    corpus_chunks_total: int
    corpus_last_updated: Optional[datetime]
    corpus_completion_pct: float
    suite_passing: int
    test_loc_ratio_avg: float


# ─────────────────────────────────────────────────────────────────
# Response schema (full settings GET)
# ─────────────────────────────────────────────────────────────────


class AdminSettingsResponse(BaseModel):
    """Response completo GET /api/v1/admin/settings/."""

    model_config = ConfigDict(from_attributes=True)

    id: UUID
    branding: BrandingSettings
    notifications: NotificationsSettings
    smtp: SmtpSettings
    general: GeneralSettings
    analytics_prefs: AnalyticsPrefs
    fiscal: FiscalSettings
    created_at: datetime
    updated_at: Optional[datetime]

File: frontend/lib/admin-settings/schemas.ts
/**
 * Zod schemas espejo de backend Pydantic AdminSettings.
 *
 * Source-of-truth contract: backend/app/admin_settings/schemas.py.
 * Drift mitigation: tests E2E (4.B.5) validan response shape vs estos
 * schemas. Si backend cambia, tests fallan visible.
 */
import { z } from "zod";

// ---------------------------------------------------------
// Branding
// ---------------------------------------------------------

export const brandingSchema = z
  .object({
    logo_url: z.string().max(500).optional().nullable(),
    primary_color: z
      .string()
      .regex(/^#[0-9A-Fa-f]{6}$/, "Color hex format #RRGGBB")
      .optional()
      .nullable(),
    secondary_color: z
      .string()
      .regex(/^#[0-9A-Fa-f]{6}$/, "Color hex format #RRGGBB")
      .optional()
      .nullable(),
    footer_text: z.string().max(500).optional().nullable(),
  })
  .strict();

export type BrandingSettings = z.infer<typeof brandingSchema>;

// ---------------------------------------------------------
// Magic Link Per Purpose Override
// ---------------------------------------------------------

export const magicLinkPurposeOverrideSchema = z
  .object({
    enabled: z.boolean().default(true),
    default_recipient_email: z.string().email().optional().nullable(),
    cc_emails: z.array(z.string().email()).optional().nullable(),
    template_id: z.string().optional().nullable(),
  })
  .strict();

export type MagicLinkPurposeOverride = z.infer<
  typeof magicLinkPurposeOverrideSchema
>;

// ---------------------------------------------------------
// Notifications
// ---------------------------------------------------------

export const notificationsSchema = z
  .object({
    client_messages_forward_enabled: z.boolean().default(true),
    client_messages_forward_to: z.string().email().optional().nullable(),
    smtp_custom: z.record(z.unknown()).optional().nullable(),
    digest_enabled: z.boolean().optional().default(false),
    digest_time_local: z
      .string()
      .regex(/^([01]\d|2[0-3]):[0-5]\d$/, "Format HH:MM")
      .default("08:00"),
    magic_link_default_sender: z.string().email().optional().nullable(),
    // #26 · WhatsApp de Marcos (E.164) editable desde el panel (antes solo env).
    marcos_whatsapp_number: z
      .string()
      .regex(/^\+?[1-9]\d{6,14}$/, "Número E.164, ej. +34637165328")
      .optional()
      .nullable(),
    magic_link_per_purpose_overrides: z
      .record(magicLinkPurposeOverrideSchema)
      .optional()
      .nullable(),
  })
  .strict();

export type NotificationsSettings = z.infer<typeof notificationsSchema>;

// ---------------------------------------------------------
// SMTP (plan v4.2 tarea 4.20 literal: 4 fields editables)
// ---------------------------------------------------------

export const smtpSchema = z
  .object({
    host: z.string().max(255).optional().nullable(),
    port: z.number().int().min(1).max(65535).optional().nullable(),
    username: z.string().max(255).optional().nullable(),
    password: z.string().max(500).optional().nullable(),
  })
  .strict();

export type SmtpSettings = z.infer<typeof smtpSchema>;

// ---------------------------------------------------------
// General
// ---------------------------------------------------------

export const generalSchema = z
  .object({
    timezone: z.string().default("Europe/Madrid"),
    locale: z
      .string()
      .regex(/^[a-z]{2}-[A-Z]{2}$/, "Format ll-CC")
      .default("es-ES"),
    date_format: z.string().default("DD/MM/YYYY"),
  })
  .strict();

export type GeneralSettings = z.infer<typeof generalSchema>;

// ---------------------------------------------------------
// Analytics Prefs
// ---------------------------------------------------------

export const analyticsPrefsSchema = z
  .object({
    show_corpus_metric: z.boolean().default(true),
  })
  .strict();

export type AnalyticsPrefs = z.infer<typeof analyticsPrefsSchema>;

// ---------------------------------------------------------
// Fiscal (identidad fiscal del emisor/consultor · punto #44)
// ---------------------------------------------------------

export const fiscalSchema = z
  .object({
    nif: z.string().max(20).optional().nullable(),
    nombre_fiscal: z.string().max(200).optional().nullable(),
    nombre_comercial: z.string().max(200).optional().nullable(),
    tipo_persona: z.enum(["F", "J"]).default("F"),
    domicilio_via: z.string().max(250).optional().nullable(),
    domicilio_cp: z.string().max(10).optional().nullable(),
    domicilio_municipio: z.string().max(120).optional().nullable(),
    domicilio_provincia: z.string().max(120).optional().nullable(),
    domicilio_pais: z.string().max(80).default("España"),
    iva_pct: z.number().min(0).max(100).default(21),
    sujeto_irpf: z.boolean().default(true),
    irpf_pct: z.number().min(0).max(100).default(15),
    iban: z.string().max(34).optional().nullable(),
    bank_holder: z.string().max(200).optional().nullable(),
    bank_institution: z.string().max(120).optional().nullable(),
    bank_bic: z.string().max(11).optional().nullable(),
  })
  .strict();

export type FiscalSettings = z.infer<typeof fiscalSchema>;

// ---------------------------------------------------------
// AdminSettings Response (full)
// ---------------------------------------------------------

export const adminSettingsResponseSchema = z.object({
  id: z.string().uuid(),
  branding: brandingSchema,
  notifications: notificationsSchema,
  smtp: smtpSchema,
  general: generalSchema,
  analytics_prefs: analyticsPrefsSchema,
  fiscal: fiscalSchema,
  created_at: z.string().datetime(),
  updated_at: z.string().datetime().nullable(),
});

export type AdminSettingsResponse = z.infer<typeof adminSettingsResponseSchema>;

// ---------------------------------------------------------
// About (read-only stats)
// ---------------------------------------------------------

export const adminSettingsAboutSchema = z.object({
  version: z.string(),
  commit_hash: z.string(),
  uptime_days: z.number(),
  active_clients_count: z.number(),
  corpus_sources_count: z.number(),
  corpus_sources_target: z.number(),
  corpus_chunks_total: z.number(),
  corpus_last_updated: z.string().datetime().nullable(),
  corpus_completion_pct: z.number(),
  // S24: suite_passing / test_loc_ratio_avg eliminados (vanity metrics fabricadas).
});

export type AdminSettingsAbout = z.infer<typeof adminSettingsAboutSchema>;

// ---------------------------------------------------------
// SMTP Test (commit fb68d16)
// ---------------------------------------------------------

export const smtpTestRequestSchema = z
  .object({
    use_admin_settings: z.boolean().default(true),
    override_host: z.string().max(255).optional().nullable(),
    override_port: z.number().int().min(1).max(65535).optional().nullable(),
    override_username: z.string().max(255).optional().nullable(),
    override_password: z.string().max(500).optional().nullable(),
  })
  .strict();

export type SmtpTestRequest = z.infer<typeof smtpTestRequestSchema>;

export const smtpTestResponseSchema = z.object({
  ok: z.boolean(),
  sent_to: z.string(),
  backend_used: z.string(),
  error: z.string().optional().nullable(),
});

export type SmtpTestResponse = z.infer<typeof smtpTestResponseSchema>;
